core/trino-main/src/main/java/io/trino/server/ThrowableMapper.java

```diff
@@ -54,6 +55,11 @@ public Response toResponse(Throwable throwable)
 
         ResponseBuilder responseBuilder = Response.serverError()
                 .header(CONTENT_TYPE, TEXT_PLAIN);
+
+        if (!Strings.isNullOrEmpty(throwable.getMessage())) {
+            responseBuilder.header("X-Trino-Error-Message", throwable.getMessage());
```

Comment:
You need to use `io.trino.client.ProtocolHeaders`. And possibly define the new header there. Also I think it should be mentioned that it is for UI related errors. As for query execution we are using `io.trino.server.QueryResource#getQueryInfo`.